src/prommis/nanofiltration/diafiltration.py

```diff
@@ -0,0 +1,774 @@
+        costing_method=DiafiltrationCostingData.cost_pump,
+        costing_method_arguments={
+            "inlet_pressure": m.atmospheric_pressure,
+            "outlet_pressure": m.operating_pressure,
+            "inlet_vol_flow": m.fs.stage3.retentate_inlet.flow_vol[0],  # diafiltrate
+        },
+    )
+    m.fs.costing.cost_process()
+
+
+def unfix_opt_variables(m):
+    """
+    Method to unfix variables for performing optimization with DOF>0
+
+    Args:
+        m: Pyomo model
+    """
+    m.fs.stage1.length.unfix()
+    m.fs.stage2.length.unfix()
+    m.fs.stage3.length.unfix()
+
+
+def add_product_constraints(
+    m,
+    Li_recovery_bound,
+    Co_recovery_bound,
+    recovery=True,
+    Li_purity_bound=None,
+    Co_purity_bound=None,
+    purity=False,
+):
+    """
+    Method to add recovery/purity constraints to the flowsheet for performing optimization
+
+    Args:
+        m: Pyomo model
+    """
+    if recovery:
+
+        @m.Constraint()
+        def Li_recovery_constraint(m):
+            return m.Li_recovery >= Li_recovery_bound
+
+        @m.Constraint()
+        def Co_recovery_constraint(m):
+            return m.Co_recovery >= Co_recovery_bound
+
+    if purity:
+        if Li_purity_bound == None:
+            raise ValueError("A lithiunm product purity bound was not provided")
+
+        if Co_purity_bound == None:
+            raise ValueError("A cobalt product purity bound was not provided")
+
+        @m.Constraint()
+        def Li_purity_constraint(m):
+            return m.Li_purity >= Li_purity_bound
+
+        @m.Constraint()
+        def Co_purity_constraint(m):
+            return m.Co_purity >= Co_purity_bound
+
+
+def add_objective(m):
+    """
+    Method to add cost objective to flowsheet for performing optimization
+
+    Args:
+        m: Pyomo model
+    """
+
+    def cost_obj(m):
+        return m.fs.costing.total_annualized_cost
+
+    m.cost_objecticve = Objective(rule=cost_obj)
+
+
+def set_scaling(m):
+    """
+    Apply scaling factors to certain constraints to improve solver performance
+
+    Args:
+        m: Pyomo model
+    """
+    m.scaling_factor = Suffix(direction=Suffix.EXPORT)
+
+    m.scaling_factor[m.fs.stage3.retentate_material_balance[0.0, 9, "Co"]] = 1e-8
+    m.scaling_factor[m.fs.stage3.retentate_material_balance[0.0, 10, "Co"]] = 1e-8
+
+
+def print_information(m):
+    """
+    Prints relevant information after solving the model
+
+    Args:
+        m: Pyomo model
+    """
+    print(
+        f"The lithium recovery is {value(m.Li_recovery) * 100}% at purity {value(m.Li_purity) * 100}"
+    )
+    print(
+        f"The cobalt recovery is {value(m.Co_recovery) * 100}% at purity {value(m.Co_purity) * 100}"
+    )
+
+    print("\ntotal membrane area")
+    print(f"{value(m.membrane_length)*value(m.w)} m2")
+
+    print("\ntotal annualized cost")
+    print(f"${value(m.fs.costing.total_annualized_cost)}")
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
This would become
```
m = ConcreteModel()
m.fs = FlowsheetBlock(dynamic=False)
m.fs.costing = DiafiltrationCosting()
main(m)
```

The test file would do something similar.